app/assets/src/components/visualizations/heatmap/Heatmap.js

```diff
@@ -400,6 +439,165 @@ export default class Heatmap {
+  }
+
+  pan(deltaX, deltaY) {
+    // Define the scrolling boundaries for the svg.
+    // Upper limits are determined by the difference between the container and svg sizes,
+    // scaled by the zoom factor.
+    let containerWidth = this.container[0][0].offsetWidth,
+      containerHeight = this.container[0][0].offsetHeight;
+    let xScrollMax =
+        (containerWidth - this.svg.attr("width")) / this.options.zoom,
+      yScrollMax =
+        (containerHeight - this.svg.attr("height")) / this.options.zoom;
+
+    // Translating the svg:
+    let gCurrentTranslate = d3.transform(this.g.attr("transform")).translate;
+    // Limit translation by the boundaries set for the svg.
+    let dx = Math.min(
+      this.options.marginLeft,
+      Math.max(deltaX + gCurrentTranslate[0], xScrollMax)
+    );
+    let dy = Math.min(
+      this.options.marginTop,
+      Math.max(deltaY + gCurrentTranslate[1], yScrollMax)
+    );
+    this.g.attr("transform", `translate(${[dx, dy]})`);
+
+    // Translating the row labels in the opposite x direction of the svg.
+    let rowLabelsCurrent = d3.transform(this.gRowLabels.attr("transform"))
+      .translate;
+    let labelsDx = Math.max(
+      0,
+      Math.min(
+        rowLabelsCurrent[0] - deltaX,
+        -xScrollMax + this.options.marginLeft
+      )
+    );
+    this.gRowLabels.attr(
+      "transform",
+      `translate(${[labelsDx, rowLabelsCurrent[1]]})`
+    );
+
+    // Translating the metadata labels in the opposite x direction of the svg (same as row labels).
+    let applyFormat = nodes => {
+      nodes.attr("transform", (d, idx) => {
+        const xOffset =
+          d.value === this.columnMetadataSortField
+            ? -this.options.metadataSortIconSize - this.options.spacing
+            : 0;
+        return `translate(${labelsDx + xOffset},
+            ${idx * this.options.minCellHeight})`;
+      });
+    };
+    let columnMetadataLabels = this.gColumnMetadata
+      .selectAll(`.${cs.columnMetadataLabel}`)
+      .data(this.options.columnMetadata, d => d.value);
+    applyFormat(columnMetadataLabels);
+
+    // Translate the "Add Metadata" link (same as row labels).
+    let metadataAddLink = this.gColumnMetadata.select(
+      `.${cs.columnMetadataAdd}`
+    );
+    let metadataAddLinkCurrent = d3.transform(metadataAddLink.attr("transform"))
+      .translate;
+    metadataAddLink.attr(
+      "transform",
+      `translate(${[labelsDx, metadataAddLinkCurrent[1]]})`
+    );
+    // Resize the add metadata line.
+    metadataAddLink
+      .select(`.${cs.metadataAddLine}`)
+      .attr(
+        "x2",
+        this.rowLabelsWidth +
+          this.columnLabels.length * this.cell.width -
+          labelsDx
+      );
+
+    // Translating the column labels in the opposite y direction of the svg.
+    let columnLabelsCurrent = d3.transform(this.gColumnLabels.attr("transform"))
+      .translate;
+    let labelsDy = Math.max(
+      this.columnLabelsHeight,
```

Comment:
Could you use `clamp` from `lodash/fp` instead of `Math.max(Math.min`? I find it's a little easier to tell what the upper and lower bound are with the `clamp` function, so it makes it more readable. Especially since you use min/max quite a bit here.